rfcs/005-group-stats.md

```diff
@@ -0,0 +1,79 @@
+# Summary
+A page at the course-instance level for group statistics. 
+
+# Basic Example Workflow
+An instructor with appropriate access uploads a reference (i.e. desired) groups assignment for their class. PrairieLearn will display group-level statistics across assessments in a given course instance. Instructor can then decide to adjust filters and parameters to get desired statistics and be able to recompute and download. 
```

Comment:
The uploaded file is the intended course group assignments to reference against for the rest of the statistics (since instructors might assign groups for their course outside of PrairieLearn, but not want to restrict students in the actual PrairieLearn assessment(s)). It's to establish some truth on what the instructor intends the groups to be. 
The CSV would be the same as in for the "upload" option in ```pages/instructorAssessmentGroups``` with the format being ```groupName,UID```.
If this offers a bit more clarity I can update the file appropriately!

Reply:
I'm still not fully understanding this. What's the interplay between this file and the *actual* group assignments that are stored in PrairieLearn?

> It's to establish some truth on what the instructor intends the groups to be.

Why do we need a separate source of truth for this? PrairieLearn already knows what all the group assignments were.

Reply:
The intention is not what the group assignments were as stored in PrairieLearn, but what they're _intended_ to be. Instructors may assign students group in their class but leave students the ability to create and join their own groups in PrairieLearn. So this addresses issues such as Student A is supposed to work with Group1, but they joined a Group5 for an assessment or Student B didn't join a group at all because they weren't in class and the instructor allows a number of group assignments to be completed alone.

Reply:
Okay, perfect, I'll add some updates to the RFC accordingly once we get the rest of this clarified! 

I was thinking that these statistics could still use this file at the very least as a filter if instructors are wanting to check if their students are working well with the groups they're supposed to be working with (e.g. "Submission Count" would be nice to do with respect to the intended group and filter out any submissions from assignments a student completed by themselves). 

I think the scenario I had in mind is slightly different, so maybe that will clarify my thinking. As you pointed out, semester-long groups are common especially for a discussion/lab section or project-based classes, so leaving that as pre-assigned and not needing a designated new file makes sense, so I think we're on the same page on that. Group assignments for in-class assessments during lecture might be different than semester-long groups where a group does everything together in a discussion or lab section, and instructors might change the assignments halfway through the semester (giving students the option to work with a new group). However, in lectures, instructors might want more flexibility/forgiveness in their group work assessments and policies. For example, going back to "Student A is supposed to work with Group1, but they joined a Group5 for an assessment", perhaps Student A joined Group 5 because none of their group showed up to class (not uncommon, particularly for big lectures) so the instructor had them join another group for that assessment. It's not so much "encouraged (but not required)" group assignments as policy, but instructors writing the assessments having in mind things such as still allowing those who show up to get the group work experience even if the rest of their group does not show and attendance policy (going back to Student B didn't join a group at all because they weren't in class and the instructor allows a number of group assignments to be completed alone). These both would force most group assessments to allow students to join and create any group and have a minimum group size of 1. Also instructors might change the group assignments halfway through the semester (or give students the chance to change groups if they want). This is why I had in mind uploading the file and having means of filtering or adjusting parameters to reflect what information/policies instructors want, but I want the same of having this be broadly useful and make sense in PL, so I'm up for whatever makes the most sense.

Reply:
I'm still open to exploring an implementation of this, but I'd like this RFC to contain a lot more information about how this will actually work. I still have a variety of open questions:

- How do you uniquely identify a group? If students can create their own groups in lecture, they're unlikely to use exactly the naming scheme that instructors do for the CSV they upload.
- How does this work when group names vary from assessment to assessment? I think this is probably related to the preceding question. From looking at the example output table, it looks like this assumes globally-consistent group names which is not likely to occur in practice.

I think implementing the other two insights you propose doesn't require us to deal with uploading this "desired groups" CSV. If you're willing to forgo the desired groups insight, this overall task should become much simpler, and IMO wouldn't even require an RFC.